aten/src/ATen/native/hammerblade/Copy.cpp

```diff
@@ -4,16 +4,32 @@
 #include <ATen/hammerblade/HammerBladeContext.h>
 #include <ATen/native/Copy.h>
 #include <ATen/native/TensorIterator.h>
+#include <ATen/native/hammerblade/Offload.h>
 
 namespace at {
 namespace native {
 
-/* These are FAKE copy kernels
- * They are just using memcpy
- */
+static bool copy_requires_temporaries(TensorIterator& iter) {
+  Device dst_device = iter.device(0);
+  Device src_device = iter.device(1);
+
+  if (dst_device == src_device) {
+    // We never require temporaries for copies on HammerBlade
+    // XXX: Be careful with this ...
+    TORCH_INTERNAL_ASSERT(dst_device.is_hammerblade(), src_device.is_hammerblade());
+    return false;
+  }
+
+  bool same_dtype = iter.dtype(0) == iter.dtype(1);
+  if (same_dtype && iter.is_contiguous()) {
+    // Contiguous same-dtype copies can always use memcpy
+    return false;
+  }
+  return true;
+}
 
 static void copy_hb_to_hb(TensorIterator& iter, bool non_blocking) {
```

Comment:
Why not overload the function `offload_memcpy` to call `tensorlib_copy_hb_to_hb` when passed iterators. Or, totally remove `offload_memcpy` and keep this? Two functions doing the same thing make API little confusing...

Reply:
Yeah I can totally see what you are saying. I like the overloading idea. Making changes now ... but it will be in the merge branch, not here ;)